core/src/main/java/org/opensearch/sql/planner/physical/UnnestOperator.java

```diff
@@ -0,0 +1,291 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import org.apache.commons.lang3.StringUtils;
+import org.opensearch.sql.data.model.ExprCollectionValue;
+import org.opensearch.sql.data.model.ExprTupleValue;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.expression.ReferenceExpression;
+
+@EqualsAndHashCode(callSuper = false)
+public class UnnestOperator extends PhysicalPlan {
+  @Getter
+  private final PhysicalPlan input;
+  @Getter
+  private final Set<String> fields; // Needs to be a Set to match legacy implementation
+  @Getter
+  private final Map<String, List<String>> groupedPathsAndFields;
+  List<Map<String, ExprValue>> result = new ArrayList<>();
+  List<String> nonNestedFields = new ArrayList<>();
+  @EqualsAndHashCode.Exclude
+  private ListIterator<Map<String, ExprValue>> flattenedResult = result.listIterator();
+
+  /**
+   * Constructor for UnnestOperator with list of map as arg.
+   * @param input : PhysicalPlan input.
+   * @param fields : List of all fields and paths for nested fields.
+   */
+  public UnnestOperator(PhysicalPlan input, List<Map<String, ReferenceExpression>> fields) {
+    this.input = input;
+    this.fields = fields.stream()
+        .map(m -> m.get("field").toString())
+        .collect(Collectors.toSet());
+    this.groupedPathsAndFields = fields.stream().collect(
+        Collectors.groupingBy(
+            m -> m.get("path").toString(),
+            mapping(
+                m -> m.get("field").toString(),
+                toList()
+            )
+        )
+    );
+  }
+
+  /**
+   * Constructor for UnnestOperator with Set of fields.
+   * @param input : PhysicalPlan input.
+   * @param fields : List of all fields for nested fields.
+   * @param groupedPathsAndFields : Map of fields grouped by their path.
+   */
+  public UnnestOperator(
+      PhysicalPlan input,
+      Set<String> fields,
+      Map<String, List<String>> groupedPathsAndFields
+  ) {
+    this.input = input;
+    this.fields = fields;
+    this.groupedPathsAndFields = groupedPathsAndFields;
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitUnnest(this, context);
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return input.hasNext() || flattenedResult.hasNext();
+  }
+
+
+  @Override
+  public ExprValue next() {
+    if (!flattenedResult.hasNext()) {
+      result.clear();
+      nonNestedFields.clear();
+
+      ExprValue inputValue = input.next();
+      generateNonNestedFieldsMap(inputValue);
+      for (String field : fields) {
+        result = flatten(field, inputValue, result, true);
+      }
+
+      for (String nonNestedField : nonNestedFields) {
+        result = flatten(nonNestedField, inputValue, result, false);
+      }
```

Comment:
Is it safe to overwrite `result` in the loop? Maybe leave a comment

Reply:
Added comments for each `flatten` function. Do you see something that isn't safe?